drivers/crypto/crypto_mchp_sha_g1.c

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright (C) 2025 Microchip Technology Inc. and its subsidiaries
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#define DT_DRV_COMPAT microchip_sha_g1_crypto
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(cypto_mchp_sha_g1);
+
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/mchp_sam_pmc.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/util.h>
+
+#define MCHP_SHA_CAPS_SUPPORT (CAP_INPLACE_OPS | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS)
+
+#define FIRST_WORD_4_PADDING_EMPTY_MSG 0x000000080
+
+struct crypto_mchp_sha_cfg {
+	sha_registers_t *const regs;
+	const struct sam_clk_cfg clock_cfg;
+};
+
+struct crypto_mchp_sha_data {
+	struct k_sem device_sem;
+};
+
+struct crypto_mchp_sha_algo_cfg {
+	enum hash_algo algo;
+	uint32_t sha_mr_algo;
+	uint8_t dgst_len;
+	uint8_t block_size;
+};
+
+struct crypto_mchp_sha_session {
+	struct crypto_mchp_sha_algo_cfg const *algo_cfg;
+	bool in_use;
+};
+
+static const struct crypto_mchp_sha_algo_cfg mchp_sha_algo_cfgs[4] = {
+	{CRYPTO_HASH_ALGO_SHA224, SHA_MR_ALGO_SHA224, 224 / BITS_PER_BYTE, 64},
+	{CRYPTO_HASH_ALGO_SHA256, SHA_MR_ALGO_SHA256, 256 / BITS_PER_BYTE, 64},
+	{CRYPTO_HASH_ALGO_SHA384, SHA_MR_ALGO_SHA384, 384 / BITS_PER_BYTE, 128},
+	{CRYPTO_HASH_ALGO_SHA512, SHA_MR_ALGO_SHA512, 512 / BITS_PER_BYTE, 128},
+};
+
+static struct crypto_mchp_sha_session mchp_sha_sessions[2];
+static struct k_sem mchp_sha_session_sem;
+
+static inline void mchp_sha_set_input(sha_registers_t *const regs, const unsigned char *data,
+				      int len)
+{
+	uint32_t i = 0, v;
+
+	while (len >= sizeof(uint32_t)) {
+		v = data[i * sizeof(uint32_t) + 3];
+		v = (v << BITS_PER_BYTE) | data[i * sizeof(uint32_t) + 2];
+		v = (v << BITS_PER_BYTE) | data[i * sizeof(uint32_t) + 1];
+		v = (v << BITS_PER_BYTE) | data[i * sizeof(uint32_t) + 0];
```

Comment:
Here it seems to me that you are just filling an `uint32_t` with 4 `uint8_t`. I think you can get the same by declaring a new pointer like:
`uint32_t *data_u32 = (uint32_t *) data;`
at the beginning of the function and then just setting:
`v = data_u32[i]`
in the while loop below. Wdyt?
In case you need to change data endianness you can use function from `zephyr/include/zephyr/sys/byteorder.h` instead of reimplementing them.

The same holds also for `mchp_sha_get_output` below.